linker: Refactoring and fixes to native library linking

petrochenkov opened the pull request:
This PR contains a bunch of code cleanup and comment rearrangements + 2 fixes for `-Zpacked-bundled-libs`.
It's better to look at individual commits.

petrochenkov committed:
linker: Factor out native library linking to a separate function

petrochenkov committed:
linker: Simplify linking of `compiler_builtins` and `profiler_builtins`

This also fixes linking of native libraries bundled into these crates when `-Zpacked-bundled-libs` is enabled

petrochenkov committed:
linker: Move some inner functions to the outside

Inline `fn unlib`

rust-highfive commented:
r? @nagisa

(rust-highfive has picked a reviewer for you, use r? to override)

petrochenkov commented:
cc @BelovDV

petrochenkov commented:
The failing run-make-fulldeps test needs to be updated, adjacent identical native libraries are no longer deduplicated if they come from different crates, but that's fine, it's not the [main motivation](https://github.com/rust-lang/rust/pull/84794).

bjorn3 commented on compiler/rustc_codegen_ssa/src/base.rs:
```diff
@@ -840,20 +840,30 @@ impl CrateInfo {
+            .filter(|&cnum| {
+                let link = !tcx.dep_kind(cnum).macros_only();
+                if link && tcx.is_compiler_builtins(cnum) {
+                    compiler_builtins = Some(cnum);
```
Shouldn't compiler builtins be included even if it isn't used according to name resolution?

petrochenkov replied:
`postorder_cnums` returns all dependency crates.
If compiler builtins end up only being used as a proc macro dependency somehow (`macros_only`), then they shouldn't get into `used_crates` and be linked.
`CrateInfo::compiler_builtins` field is also only used twice - in `fn ignored_for_lto` and `fn add_upstream_rust_crates`, which are not for proc macro dependencies.

bjorn3 replied:
I thought compiler_builtins was automatically injected for all crates except libcore and I assumed that it wouldn't be listed as regular dependency. I can't find the relevant code though, so maybe I was wrong.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2514,23 +2496,6 @@ fn add_upstream_rust_crates<'a>(
-            cnum,
-            &Default::default(),
-        );
-    }
```
This has to be placed after everything, not just after rust crates but also after native dependencies in case libgcc_s or libcompiler_rt aren't linked.

petrochenkov replied:
Compiler builtins are placed after all crates and all native libs those crates bring.
It's not something this PR changes, it's the behavior both before and after this PR.

bjorn3 replied:
I can't find the code that causes compiler_builtins to be placed after all native libs.

bjorn3 replied:
What if libgcc_s.so is linked? It should come before compiler_builtins to ensure that the intrinsics resolve in libgcc_s.so rather than libcompiler_builtins.rlib I think as the former is much more optimized.

petrochenkov replied:
Also, is compiler-rt optimized on par with libgcc (I would expect so)?
If compiler_builtins is built with enabling C components borrowed from compiler-rt (https://docs.rs/crate/compiler_builtins/0.1.83/features#c, IIRC) then compiler_builtins should also be good in this regard.

bjorn3 replied:
> But right now it's linked last, so I guess the versions of intrinsics from compiler_builtins are typically used on GNU targets (unless crt-static is enabled or something like that).

Right, I confirmed this just now. I was under the impression that compiler_builtins would be after dynamic libraries, but because this is not the case I will resolve this conversation.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2259,12 +2259,12 @@ fn add_native_libs_from_crate(
+    // will provide them to the linker itself.
+    // The option doesn't currently apply to native libraries bundled into rlibs, regardless
+    // of the value of `-Zpacked-bundled-libs`, this may be changed in the future.
+    let link_native = sess.opts.unstable_opts.link_native_libraries;
```
This commit goes against the purpose of `-Zlink-native-libraries`, doesn't it?

petrochenkov replied:
Maybe, but I'm not sure.
If a crate brings some private C code with it, compiles it, and then bundles it, does it concern the outer build system?
In case of native libs bundled in the old way (as object files) we are not "unbundling" them if `-Zlink-native-libraries=false` is used, this commit at least makes the new behavior consistent with that.
`-Zpacked-bundled-libs` is unstable, so all of this is up to discussion.

bjorn3 replied:
I think `-Zlink-native-libraries` only makes sense if the build system invoking rustc is also the one invoking the linker. As such this build system would know about the native lib bundled with the rlib. If these build systems are not the same, it is almost guaranteed to break as crates can add dependencies that the linking build system doesn't know about.

petrochenkov replied:
I think I'll remove this commit after all.
Libraries bundled in the old way are added as object files so some third party build system linking rlibs as classic staticlibs will do more or less the right thing implicitly.
But libraries bundled in the old way will be ignored and not added to the link if rlibs are treated as staticlibs, so the build system will need to do something to get them linked anyway.

petrochenkov replied:
I've removed this change.

klensy commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2667,6 +2535,132 @@ fn add_upstream_native_libraries(
+        return;
+    }
+
+    let dst = tmpdir.join(cratepath.file_name().unwrap());
+    let name = cratepath.file_name().unwrap().to_str().unwrap();
+    let name = &name[3..name.len() - 5]; // chop off lib/.rlib
```
Maybe add here some assert that we actually get expected extension, not just cut something?

bjorn3 replied:
Currently the code is correct as the metadata reader mandates lib and .rlib as prefix and suffix.

bors commented:
:umbrella: The latest upstream changes (presumably #103978) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov committed:
linker: Support mixing crates built with different values of `-Zpacked_bundled_libs`

So you can change the value of `-Zpacked_bundled_libs` without rebuilding standard library

petrochenkov committed:
Update run-make-fulldeps tests

Adjacent identical native libraries are no longer deduplicated if they come from different crates

nagisa commented:
r? @bjorn3 seems to be way more on-top of this than I am.

petrochenkov commented:
Updated with the change from https://github.com/rust-lang/rust/pull/103311#discussion_r1001690467 removed.
@rustbot ready

bjorn3 commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit edb9175eb4e0641138f9d90d4877f86469b3a950 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 edb9175eb4e0641138f9d90d4877f86469b3a950 -->
<!-- homu: {"type":"Approved","sha":"edb9175eb4e0641138f9d90d4877f86469b3a950","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit edb9175eb4e0641138f9d90d4877f86469b3a950 with merge e1392304fca9da9c1e1a81251f0384456380812c...
<!-- homu: {"type":"BuildStarted","head_sha":"edb9175eb4e0641138f9d90d4877f86469b3a950","merge_sha":"e1392304fca9da9c1e1a81251f0384456380812c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3451217903/jobs/5760466786)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3451217903/jobs/5760466786","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-stable`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3451217903/jobs/5760297302) [(plain)](https://github.com/rust-lang-ci/rust/commit/e1392304fca9da9c1e1a81251f0384456380812c/checks/9450376958/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
# We don't compile `opaque` with either optimizations or instrumentation.
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights  -Clink-args=-fuse-ld=gold opaque.rs || exit 1
# Compile the test program with instrumentation
mkdir -p "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/prof_data_dir" || exit 1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights  -Clink-args=-fuse-ld=gold interesting.rs \
 -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/prof_data_dir" -O -Ccodegen-units=1 || exit 1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights  -Clink-args=-fuse-ld=gold main.rs -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/prof_data_dir" -O || exit 1
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/rustcc83VLM/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.1.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.10.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.11.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.2.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.3.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.4.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.5.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.6.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.7.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.8.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.9.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.4g86d6gt786od2ny.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/libinteresting.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/libopaque.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-31b5cc48820ae0df.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-245e68d84dc0309a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d2ea0a854fda672c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-24562481dedfd471.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-194a45968167b49e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-09596d3041f17fd8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-181533669a2516a5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-14b388f38119c258.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-87e60e5d00044771.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2ba187f11260889c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-3f2f59c88ec20cee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-824cd1ee1272e83c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-e280e795be985f14.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-a5802649cd08021d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-628eac137706518b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-8169d49e60c0ddd2.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a2085980cc14b4fa.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0ae6dfd9f53e8a76.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "-u" "__llvm_profile_runtime" "-fuse-ld=gold"
  = note: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-branch-weights/pgo-branch-weights/main.main.cbcd4161-cgu.3.rcgu.o:main.cbcd4161-cgu.3:function std::sys_common::backtrace::__rust_begin_short_backtrace: error: undefined reference to '__llvm_profile_instrument_target'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:24: all] Error 1
------------------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen  -g -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen" test.rs
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/rustcrOnMc2/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.1.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.2.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.3.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.4.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.2k7bs2fd4dv6n3lm.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-31b5cc48820ae0df.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-245e68d84dc0309a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d2ea0a854fda672c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-24562481dedfd471.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-194a45968167b49e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-09596d3041f17fd8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-181533669a2516a5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-14b388f38119c258.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-87e60e5d00044771.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2ba187f11260889c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-3f2f59c88ec20cee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-824cd1ee1272e83c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-e280e795be985f14.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-a5802649cd08021d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-628eac137706518b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-8169d49e60c0ddd2.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a2085980cc14b4fa.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0ae6dfd9f53e8a76.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs" "-u" "__llvm_profile_runtime"
  = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen/pgo-gen/test.test.b2d137c1-cgu.0.rcgu.o: in function `core::ops::function::FnOnce::call_once':
          /checkout/library/core/src/ops/function.rs:510: undefined reference to `__llvm_profile_instrument_target'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:12: all] Error 1
------------------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
# Compile the test program with instrumentation
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use  -Copt-level=2 -Ccodegen-units=1 -Cllvm-args=-disable-preinline -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use" main.rs
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/rustcxdC0Gz/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/main.main.cbcd4161-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/main.4g86d6gt786od2ny.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-31b5cc48820ae0df.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-245e68d84dc0309a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d2ea0a854fda672c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-24562481dedfd471.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-194a45968167b49e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-09596d3041f17fd8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-181533669a2516a5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-14b388f38119c258.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-87e60e5d00044771.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2ba187f11260889c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-3f2f59c88ec20cee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-824cd1ee1272e83c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-e280e795be985f14.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-a5802649cd08021d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-628eac137706518b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-8169d49e60c0ddd2.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a2085980cc14b4fa.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0ae6dfd9f53e8a76.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/main" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "-u" "__llvm_profile_runtime"
  = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/main.main.cbcd4161-cgu.0.rcgu.o: in function `std::sys_common::backtrace::__rust_begin_short_backtrace':
          main.cbcd4161-cgu.0:(.text._ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hcd72613ad097f94dE+0x1c): undefined reference to `__llvm_profile_instrument_target'
          /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-use/pgo-use/main.main.cbcd4161-cgu.0.rcgu.o: in function `core::option::Option<T>::map_or_else':
          main.cbcd4161-cgu.0:(.text._ZN4core6option15Option$LT$T$GT$11map_or_else17hcb819f086615ff69E+0x201): undefined reference to `__llvm_profile_instrument_memop'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:31: all] Error 1
------------------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
# We don't compile `opaque` with either optimizations or instrumentation.
# We don't compile `opaque` with either optimizations or instrumentation.
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion   opaque.rs
# Compile the test program with instrumentation
mkdir -p "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion"/prof_data_dir
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion   interesting.rs \
 -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion"/prof_data_dir -O -Ccodegen-units=1
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion   main.rs -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion"/prof_data_dir -O
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/rustcvgKdi5/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.1.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.10.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.11.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.12.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.13.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.2.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.3.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.4.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.5.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.6.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.7.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.8.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.9.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.4g86d6gt786od2ny.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/libinteresting.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/libopaque.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-31b5cc48820ae0df.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-245e68d84dc0309a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d2ea0a854fda672c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-24562481dedfd471.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-194a45968167b49e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-09596d3041f17fd8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-181533669a2516a5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-14b388f38119c258.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-87e60e5d00044771.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2ba187f11260889c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-3f2f59c88ec20cee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-824cd1ee1272e83c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-e280e795be985f14.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-a5802649cd08021d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-628eac137706518b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-8169d49e60c0ddd2.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a2085980cc14b4fa.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0ae6dfd9f53e8a76.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "-u" "__llvm_profile_runtime"
  = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/main.main.cbcd4161-cgu.3.rcgu.o: in function `std::sys_common::backtrace::__rust_begin_short_backtrace':
          main.cbcd4161-cgu.3:(.text._ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hcd72613ad097f94dE+0x15): undefined reference to `__llvm_profile_instrument_target'
          /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/libinteresting.rlib(interesting.interesting.8805ca9a-cgu.0.rcgu.o): in function `call_a_bunch_of_functions':
          interesting.8805ca9a-cgu.0:(.text.call_a_bunch_of_functions+0x3d): undefined reference to `__llvm_profile_instrument_target'
          /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-indirect-call-promotion/pgo-indirect-call-promotion/libinteresting.rlib(interesting.interesting.8805ca9a-cgu.0.rcgu.o): in function `call_a_bunch_of_trait_methods':
          interesting.8805ca9a-cgu.0:(.text.call_a_bunch_of_trait_methods+0x47): undefined reference to `__llvm_profile_instrument_target'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:16: all] Error 1
------------------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile  -g -Z profile test.rs
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/rustchzO32M/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/test.test.b2d137c1-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/test.2k7bs2fd4dv6n3lm.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-31b5cc48820ae0df.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-245e68d84dc0309a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-d2ea0a854fda672c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-24562481dedfd471.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-194a45968167b49e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-09596d3041f17fd8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-181533669a2516a5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-14b388f38119c258.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-87e60e5d00044771.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-2ba187f11260889c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-3f2f59c88ec20cee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-824cd1ee1272e83c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-e280e795be985f14.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-a5802649cd08021d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-628eac137706518b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-8169d49e60c0ddd2.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a2085980cc14b4fa.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0ae6dfd9f53e8a76.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/test" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs"
  = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/test.test.b2d137c1-cgu.0.rcgu.o: in function `__llvm_gcov_writeout':
          test.b2d137c1-cgu.0:(.text.__llvm_gcov_writeout+0x34): undefined reference to `llvm_gcda_start_file'
          /usr/bin/ld: test.b2d137c1-cgu.0:(.text.__llvm_gcov_writeout+0x80): undefined reference to `llvm_gcda_emit_function'
          /usr/bin/ld: test.b2d137c1-cgu.0:(.text.__llvm_gcov_writeout+0x9e): undefined reference to `llvm_gcda_emit_arcs'
          /usr/bin/ld: test.b2d137c1-cgu.0:(.text.__llvm_gcov_writeout+0xb6): undefined reference to `llvm_gcda_summary_info'
          /usr/bin/ld: test.b2d137c1-cgu.0:(.text.__llvm_gcov_writeout+0xbb): undefined reference to `llvm_gcda_end_file'
          /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/profile/profile/test.test.b2d137c1-cgu.0.rcgu.o: in function `__llvm_gcov_init':
          test.b2d137c1-cgu.0:(.text.__llvm_gcov_init+0x10): undefined reference to `llvm_gcov_init'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:6: all] Error 1
------------------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto  -Copt-level=3 -Clto=fat -Cprofile-generate="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto" test.rs
--- stderr -------------------------------
error: linking with `cc` failed: exit status: 1
  |
  |
  = note: "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto/rustcJloaa0/symbols.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto/test.test.b2d137c1-cgu.1.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-2e0bff6c30e10485.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto/test" "-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "-u" "__llvm_profile_runtime"
  = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/pgo-gen-lto/pgo-gen-lto/test.test.b2d137c1-cgu.1.rcgu.o: in function `std::sys_common::backtrace::__rust_begin_short_backtrace':
          test.b2d137c1-cgu.1:(.text._ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h4eecd744a6715adbE+0x15): undefined reference to `__llvm_profile_instrument_target'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
error: aborting due to previous error

make: *** [Makefile:12: all] Error 1
------------------------------------------
```

</details>

petrochenkov committed:
linker: Link `profiler_builtins` even if it's marked as `NotLinked`

petrochenkov commented:
I've added a commit to link `profiler_builtins` even if is's marked as `Linkage::NotLinked`.
It shouldn't be `NotLinked`, the fact that it's marked as such is some pre-existing issue in dependency_format.rs, so I've also added a FIXME comment.
@bors r=bjorn3

bors commented:
:pushpin: Commit c2358a15f31b35971a59c4bda138409c527fe4fc has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 c2358a15f31b35971a59c4bda138409c527fe4fc -->
<!-- homu: {"type":"Approved","sha":"c2358a15f31b35971a59c4bda138409c527fe4fc","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
It's not the nicest solution, but it should work for now.

@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #103913

bors commented:
:pushpin: Commit c2358a15f31b35971a59c4bda138409c527fe4fc has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 c2358a15f31b35971a59c4bda138409c527fe4fc -->
<!-- homu: {"type":"Approved","sha":"c2358a15f31b35971a59c4bda138409c527fe4fc","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented:
Didn't see you already posted an r+ :)

bors commented:
:hourglass: Testing commit c2358a15f31b35971a59c4bda138409c527fe4fc with merge fb6667a233fa677f6fbfa6067913ef3d32480317...
<!-- homu: {"type":"BuildStarted","head_sha":"c2358a15f31b35971a59c4bda138409c527fe4fc","merge_sha":"fb6667a233fa677f6fbfa6067913ef3d32480317"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3452955657/jobs/5764148565)
Approved by: bjorn3
Pushing fb6667a233fa677f6fbfa6067913ef3d32480317 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"bjorn3","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3452955657/jobs/5764148565"},"merge_sha":"fb6667a233fa677f6fbfa6067913ef3d32480317"} -->

rust-timer commented:
Finished benchmarking commit (fb6667a233fa677f6fbfa6067913ef3d32480317): [comparison URL](https://perf.rust-lang.org/compare.html?start=6284998a2677d7e3e8420db783f3aa4fd80d7423&end=fb6667a233fa677f6fbfa6067913ef3d32480317&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6284998a2677d7e3e8420db783f3aa4fd80d7423&end=fb6667a233fa677f6fbfa6067913ef3d32480317&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6284998a2677d7e3e8420db783f3aa4fd80d7423&end=fb6667a233fa677f6fbfa6067913ef3d32480317&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=6284998a2677d7e3e8420db783f3aa4fd80d7423&end=fb6667a233fa677f6fbfa6067913ef3d32480317&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
